### PR TITLE
UCP/PROTO: Add multi-fragment protocols infrastructure

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/ucp/proto/proto_common.h [hunk 1]
```diff
@@ -30,19 +32,45 @@ typedef struct {
 } ucp_proto_common_init_params_t;
 
 
+typedef struct {
+    ucp_lane_map_t          lane_map;
```
Comment (dmitrygx) [inline src/ucp/proto/proto_common.h:36]: could u pls describe the purpose of the fields?

Comment (dmitrygx) [reply src/ucp/proto/proto_common.h:36]: [q] why need caching the first lane?

**Comment by brminich** [comment]
coverity error
```
[2020-08-12T23:31:25.879Z] + '[' 1 -gt 0 ']'

[2020-08-12T23:31:25.879Z] + cov-format-errors --dir /scrap/jenkins/workspace/ucx-10/cov_build_devel_6608 --emacs-style

[2020-08-12T23:31:25.879Z] Error: OVERRUN:

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:83:

[2020-08-12T23:31:25.879Z] cond_true: Condition "1 /* UCS_LOG_LEVEL_TRACE <= UCS_LOG_LEVEL_TRACE_POLL */", taking true branch.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:83:

[2020-08-12T23:31:25.879Z] cond_true: Condition "UCS_LOG_LEVEL_TRACE <= ucs_global_opts.log_component.log_level", taking true branch.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:83:

[2020-08-12T23:31:25.879Z] cond_true: Condition "1 /* UCS_LOG_LEVEL_TRACE <= UCS_LOG_LEVEL_TRACE_POLL */ && UCS_LOG_LEVEL_TRACE <= ucs_global_opts.log_component.log_level", taking true branch.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:88:

[2020-08-12T23:31:25.879Z] cond_true: Condition "params->flags & UCP_PROTO_COMMON_INIT_FLAG_SEND_ZCOPY", taking true branch.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:89:

[2020-08-12T23:31:25.879Z] cond_false: Condition "select_param->dt_class == UCP_DATATYPE_GENERIC", taking false branch.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:89:

[2020-08-12T23:31:25.879Z] cond_false: Condition "select_param->dt_class == UCP_DATATYPE_IOV", taking false branch.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:96:

[2020-08-12T23:31:25.879Z] if_end: End of if statement.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:97:

[2020-08-12T23:31:25.879Z] if_fallthrough: Falling through to end of if statement.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:105:

[2020-08-12T23:31:25.879Z] if_end: End of if statement.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:107:

[2020-08-12T23:31:25.879Z] cond_true: Condition "_min_a < _min_b", taking true branch.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:110:

[2020-08-12T23:31:25.879Z] cond_true: Condition "lane_map", taking true branch.

[2020-08-12T23:31:25.879Z] path:/scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:110:

[2020-08-12T23:31:25.879Z] cond_true: Condition "lane < 64", taking true branch.

[2020-08-12T23:31:25.879Z] /scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:110:

[2020-08-12T23:31:25.879Z] cond_at_most: Checking "lane < 64" implies that "lane" may be up to 63 on the true branch.

[2020-08-12T23:31:25.879Z] /scrap/jenkins/workspace/ucx-10/src/ucp/proto/proto_common.c:112:

[2020-08-12T23:31:25.879Z] overrun-local: Overrunning array "ep_config_key->lanes" of 6 8-byte elements at element index 63 (byte offset 504) using index "lane" (which evaluates to 63).

[2020-08-12T23:31:25.879Z] + echo 'not ok 1 Coverity Detected 1 failures # http://hpc-master.lab.mtl.com:8080/job/ucx/6608//artifact/cov_build_devel_6608/output/errors/index.html'

[2020-08-12T23:31:25.879Z] + echo Coverity report: http://hpc-master.lab.mtl.com:8080/job/ucx/6608//artifact/cov_build_devel_6608/output/errors/index.html

[2020-08-12T23:31:25.879Z] Coverity report: http://hpc-master.lab.mtl.com:8080/job/ucx/6608//artifact/cov_build_devel_6608/output/errors/index.html

[2020-08-12T23:31:25.879Z] + printf '%s\t%s\n' Coverity http://hpc-master.lab.mtl.com:8080/job/ucx/6608//artifact/cov_build_devel_6608/output/errors/index.html

[2020-08-12T23:31:25.879Z] + module unload tools/cov
```

File: src/ucp/proto/proto_common.c [hunk 2]
```diff
@@ -158,6 +161,34 @@ ucp_proto_common_find_lanes(const ucp_proto_common_init_params_t *params,
...
+
+    if (!(params->flags & UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY)) {
+        recv_time.m = pack_time.m;
+    }
+
+    return recv_time;
+}
+
+static void
+ucp_proto_common_add_perf(const ucp_proto_common_init_params_t *params,
```
Comment (brminich) [inline src/ucp/proto/proto_common.c:181]: maybe pass `ucp_proto_init_params_t` instead?

File: src/ucp/proto/proto_multi.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ */
+typedef struct {
+    ucp_lane_index_t               num_lanes;  /* Number of lanes to use */
+    ucp_md_map_t                   reg_md_map; /* Memory domains to register on */
+    ucp_proto_multi_lane_priv_t    lanes[0];   /* Array of lanes */
+} ucp_proto_multi_priv_t;
+
+
+/**
+ * Initializatoin parameters for multi-lane protocol
```
Comment (brminich) [inline src/ucp/proto/proto_multi.h:45]: `Initialization`

**Pull request merged**